File: lib/data/entity/message_data_item.dart
import 'dart:convert';

import 'package:floor/floor.dart';
import 'package:test_twilio/common/extensions/string_extensions.dart';
import 'package:test_twilio/model/reaction_attribute.dart';

@Entity(tableName: "message_table", primaryKeys: ["sid", "uuid"])
class MessageDataItem {
  String? sid;
  String? conversationSid;
  String? participantSid;
  int? type;
  String? author;
  int? dateCreated;
  String? body;
  int? index;
  String? attributes;
  int? direction;
  int? sendStatus;
  String? uuid;
  String? mediaSid;
  String? mediaFileName;
  String? mediaType;
  int? mediaSize;
  String? mediaUri;
  int? mediaDownloadId;
  int? mediaDownloadedBytes;
  int? mediaDownloadState = 0;
  bool? mediaUploading = false;
  int? mediaUploadedBytes;
  String? mediaUploadUri;
  int? errorCode = 0;

  @ignore
  MessageDataItem.fromMap(Map<String, dynamic> data) {
    sid = data["sid"]?.toString();
    conversationSid = data["conversationSid"]?.toString();
    participantSid = data["participantSid"]?.toString();
    type = data["type"]?.toString().toInt();
    author = data["author"]?.toString();
    dateCreated = data["dateCreated"]?.toString().toInt();
    body = data["body"]?.toString();
    index = data["index"]?.toString().toInt();
    attributes = data["attributes"]?.toString();
    direction = data["direction"]?.toString().toInt();
    sendStatus = data["sendStatus"]?.toString().toInt();
    uuid = data["uuid"]?.toString();
    mediaSid = data["mediaSid"]?.toString();
    mediaFileName = data["mediaFileName"]?.toString();
    mediaType = data["mediaType"]?.toString();
    mediaSize = data["mediaSize"]?.toString().toInt();
    mediaUri = data["mediaUri"]?.toString();
    mediaDownloadId = data["mediaDownloadId"]?.toString().toInt();
    mediaDownloadedBytes = data["mediaDownloadedBytes"]?.toString().toInt();
    mediaDownloadState = data["mediaDownloadState"]?.toString().toInt();
    mediaUploading = data["mediaUploading"]?.toString().toBool();
    mediaUploadedBytes = data["mediaUploadedBytes"]?.toString().toInt();
    mediaUploadUri = data["mediaUploadUri"]?.toString();
    errorCode = data["errorCode"]?.toString().toInt();
  }

  @ignore
  Map<String, dynamic> toMap() {
    Map<String, dynamic> data = {};

    data["sid"] = sid;
    data["conversationSid"] = conversationSid;
    data["participantSid"] = participantSid;
    data["type"] = type;
    data["author"] = author;
    data["dateCreated"] = dateCreated;
    data["body"] = body;
    data["index"] = index;
    data["attributes"] = attributes;
    data["direction"] = direction;
    data["sendStatus"] = sendStatus;
    data["uuid"] = uuid;
    data["mediaSid"] = mediaSid;
    data["mediaFileName"] = mediaFileName;
    data["mediaType"] = mediaType;
    data["mediaSize"] = mediaSize;
    data["mediaUri"] = mediaUri;
    data["mediaDownloadId"] = mediaDownloadId;
    data["mediaDownloadedBytes"] = mediaDownloadedBytes;
    data["mediaDownloadState"] = mediaDownloadState;
    data["mediaUploading"] = mediaUploading;
    data["mediaUploadedBytes"] = mediaUploadedBytes;
    data["mediaUploadUri"] = mediaUploadUri;
    data["errorCode"] = errorCode;
    return data;
  }

  ReactionAttribute? toReactionAttribute() {
    try {
      return ReactionAttribute.fromJson(Map<String, dynamic>.from(jsonDecode(attributes!)));
    } catch (e) {
      return null;
    }
  }
}

File: android/app/src/main/kotlin/com/example/test_twilio/entity/MessageDataItem.kt
class MessageDataItem(
    val sid: String?,
    val conversationSid: String?,
    val participantSid: String?,
    var type: Int?,
    val author: String?,
    val dateCreated: Long?,
    val body: String?,
    val index: Long?,
    val attributes: String?,
    val direction: Int?,
    var sendStatus: Int?,
    val uuid: String?,
    val mediaSid: String? = null,
    val mediaFileName: String? = null,
    val mediaType: String? = null,
    val mediaSize: Long? = null,
    val mediaUri: String? = null,
    val mediaDownloadId: Long? = null,
    val mediaDownloadedBytes: Long? = null,
    val mediaDownloadState: Int? = 0,
    var mediaUploading: Boolean? = false,
    var mediaUploadedBytes: Long? = null,
    val mediaUploadUri: String? = null,
    var errorCode: Int? = 0,
    val filePath: String? = null
) {
    companion object {
        fun fromMap(data: HashMap<String, Any?>): MessageDataItem {
            return MessageDataItem(
                sid = data["sid"]?.toString(),
                conversationSid = data["conversationSid"]?.toString(),
                participantSid = data["participantSid"]?.toString(),
                type = data["type"]?.toString()?.toInt(),
                author = data["author"]?.toString(),
                dateCreated = data["dateCreated"]?.toString()?.toLong(),
                body = data["body"]?.toString(),
                index = data["index"]?.toString()?.toLong(),
                attributes = data["attributes"]?.toString(),
                direction = data["direction"]?.toString()?.toInt(),
                sendStatus = data["sendStatus"]?.toString()?.toInt(),
                uuid = data["uuid"]?.toString(),
                mediaSid = data["mediaSid"]?.toString(),
                mediaFileName = data["mediaFileName"]?.toString(),
                mediaType = data["mediaType"]?.toString(),
                mediaSize = data["mediaSize"]?.toString()?.toLong(),
                mediaUri = data["mediaUri"]?.toString(),
                mediaDownloadId = data["mediaDownloadId"]?.toString()?.toLong(),
                mediaDownloadedBytes = data["mediaDownloadedBytes"]?.toString()?.toLong(),
                mediaDownloadState = data["mediaDownloadState"]?.toString()?.toInt(),
                mediaUploading = data["mediaUploading"]?.toString()?.toBoolean(),
                mediaUploadedBytes = data["mediaUploadedBytes"]?.toString()?.toLong(),
                mediaUploadUri = data["mediaUploadUri"]?.toString(),
                errorCode = data["errorCode"]?.toString()?.toInt(),
                filePath = data["filePath"]?.toString(),
            )
        }
    }
    
    fun toMap() : HashMap<String, Any?> {
        val map = hashMapOf<String, Any?>()
        
        map["sid"] = sid
        map["conversationSid"] = conversationSid
        map["participantSid"] = participantSid
        map["type"] = type
        map["author"] = author
        map["dateCreated"] = dateCreated
        map["body"] = body
        map["index"] = index
        map["attributes"] = attributes
        map["direction"] = direction
        map["sendStatus"] = sendStatus
        map["uuid"] = uuid
        map["mediaSid"] = mediaSid
        map["mediaFileName"] = mediaFileName
        map["mediaType"] = mediaType
        map["mediaSize"] = mediaSize
        map["mediaUri"] = mediaUri
        map["mediaDownloadId"] = mediaDownloadId
        map["mediaDownloadedBytes"] = mediaDownloadedBytes
        map["mediaDownloadState"] = mediaDownloadState
        map["mediaUploading"] = mediaUploading
        map["mediaUploadedBytes"] = mediaUploadedBytes
        map["mediaUploadUri"] = mediaUploadUri
        map["errorCode"] = errorCode
        map["filePath"] = filePath
        return map
    }
}
